File: mates.js
/* Copiright Adriano Porcheddu adriano@eja.it 2022*/

//Número PI
var pi = 3.14159265358979323846;

//##################################OPERACIONES BÁSICAS

//Suma números
function suma() {
	var tot = 0;
	for (n in arguments) {
		tot += arguments[n];
	}
	return tot
}

//Resta números
function resta() {
	tot = arguments[0];
	for (i = 1; i < arguments.length; i++) {
		tot -= arguments[i];
	}
	return tot
}

//Multiplica números
function multiplica() {
	var tot = 1;
	for (n in arguments) {
		tot *= arguments[n];
	}
	return tot
}

//Divide números
function divide() {
	tot = arguments[0];
	for (i = 1; i < arguments.length; i++) {
		tot /= arguments[i];
	}
	return tot
}

//Da el resto de un número(a) entre otro(b)
function verificaResto(a, b) {
	return a%b
}

//Expone un número(a) a otro(b)
function expone(a, b) {
	return a**b
}

//Devuelve todos los números primos existentes hasta el límite(limite)
function primos(limite) {
	var primosAr = [2];
	for (var i=primosAr[0]; i < limite; i++) {
		var esPrimo = true;
		for (n in primosAr) {
			if (verificaResto(i, primosAr[n]) == 0) {
				esPrimo = false; 
			}
		} 		
		if (esPrimo) {
			primosAr.push(i); 	
		} 	
	} 	
	return primosAr
}

//Devuelve la división de el número de números primos entre sus decenas (veces) veces, un ejemplo sería: ({1, 2, 3, 5, 7} son 5) => (5/10); luego {11, 13, 17, 19} son 4 => ((4+5)/20)
function primosEntreDecenas(veces) {
	//10, 20, 30... (decenas)
	var decenas = [];
	//Fracciones finales
	var fracciones = []; 
	//Primos requeridos
	var allPrimes = primos(veces*10); 
	//5, 9, 11... (sumas de los primos por cada decena)
	var decs = []; 	
	//Ajusta las variables acorde con las veces(veces) requeridas
	for (var i = 1; i < veces+1; i++) {
		decenas.push((i*10).toString()); 
		decs.push(1);
	}
	//Calcula las fracciones
	for (var i = 0; i < decenas.length; i++) {
		for (n in allPrimes) {
			if (allPrimes[n] < decenas[i]) { 		
				for (var a = 1; a < decenas.length+1; a++) {
					if (decenas[i] == a*10) { 	
						decs[i]++;
					} 				
				} 		
			} 		
		} 	
		fracciones.push(divide(decs[i], decenas[i])); 
	}
	return fracciones
}

//Devuelve la media de un array de números
function media(nA) {
	var nAL = nA.length;
	var suma = 0;
	for (var i = 0; i < nAL; i++) {
		suma += parseFloat(nA[i]);
	}
	var res = (suma / nAL);
	return res
}

//################################GEOMETRÍA

//CÍRCULO
function perimetroCirculo(radio) {
	return (multiplica(radio, pi, 2))
}
function areaCirculo(radio) {
	return (multiplica(expone(radio, 2), pi))
}

function circulo(radio) {
	var datos = {
		"radio": radio,
		"area": areaCirculo(radio),
		"perimetro": perimetroCirculo(radio)
	};
	return datos
}

//CUADRADO
function perimetroCuadrado(lado) {
	return multiplica(lado, 4)
}

function areaCuadrado(lado) {
	return expone(lado, 2)
}

function cuadrado(lado) {
	var datos = {
		"lado": lado,
		"area": areaCuadrado(lado),
		"perimetro": perimetroCuadrado(lado)
	};
	return datos
}

//TRIÁNGULO EQUILÁTERO
function perimetroTrianguloEquilatero(lado) {
	return multiplica(lado, 3)
}

function alturaTrianguloEquilatero(lado) {
	var a, b, c;
	c = lado;
	b = divide(lado, 2);
	a = Math.sqrt(expone(c, 2) - expone(b, 2));
	return a
}

function areaTrianguloEquilatero(lado) {
	area = divide(multiplica(alturaTrianguloEquilatero(lado), lado), 2);
	return area
}

function trianguloEquilatero(lado) {
	var datos = {
		"lado": lado,
		"area": areaTrianguloEquilatero(lado),
		"perimetro": perimetroTrianguloEquilatero(lado)
	};
	return datos
}

//TRIÁNGULO RECTÁNGULO
function perimetroTrianguloRectangulo(cat1, cat2) {
	hip = hipotenusa(cat1, cat2);
	return suma(cat1, cat2, hip)
}

function hipotenusa(cat1, cat2) {
	hip = sqrt(suma(expone(cat1, 2), expone(cat2, 2)));
	return hip
}

function areaTrianguloRectangulo(cat1, cat2) {
	return divide(multiplica(cat1, cat2), 2)
}

function trianguloRectangulo(cat1, cat2) {
	var datos = {
		"cateto1": cat1,
		"cateto2": cat2,
		"hipotenusa": hipotenusa(cat1, cat2),
		"perimetro": perimetroTrianguloRectangulo(cat1, cat2),
		"area": areaTrianguloRectangulo(cat1, cat2)
	};
	return datos
}

File: mates.py
# Copyright Adriano Porcheddu adriano@eja.it 2022

import sys
from logging import debug, basicConfig, ERROR
import math

basicConfig(level=ERROR, format='%(asctime)s :: %(levelname)s :: %(funcName)s :: %(message)s')

#Número PI

pi = 3.14159265358979323846

#OPERACIONES BÁSICAS

#Suma números
def suma(*arguments): 
    tot = 0
    for n in arguments:
        tot += n 
    debug('la suma da {}'.format(tot))
    return tot

#Resta números
def resta(*arguments):
    tot = arguments[0]
    for x in range(0, len(arguments)):
        if x == 0:
            tot = arguments[x]
        else:
            tot -= arguments[x]
    debug('la resta da {}'.format(tot))
    return tot

#Multiplica números
def multiplica(*arguments):
    tot = 1
    for n in arguments:
        tot *= n
    debug('la multiplicación da {}'.format(tot))
    return tot

#Divide números
def divide(*arguments):
    tot = arguments[0]
    x = 1
    for x in arguments:
        tot /= x
    debug('la división da {}'.format(tot))
    return tot

#Da el resto de un número(a) entre otro(b)
def verifica_resto(a, b, debug=False):
    tot = a%b
    if debug:
        debug('el resto es {}'.format(tot))
    return tot

#Expone un número(a) a otro(b)
def expone(a, b):
    tot = a**b
    debug('el exponente da {}'.format(tot))
    return tot

#Devuelve todos los números primos existentes hasta el límite(limite)
def primos(limite):
    primos_ar = [2]
    i = 2
    while i<limite:
        es_primo = True
        for n in primos_ar:
            if verifica_resto(i, n) == 0:
                es_primo = False

        if es_primo:
            primos_ar.append(i)
        i += 1
    debug('los primos hasta {x} son {y}'.format(x=limite, y=primos_ar))
    return primos_ar

#Devuelve la división del número de números primos entre sus decenas (veces) veces, un ejemplo sería: ({1, 2, 3, 5, 7} son 5) => (5/10); luego {11, 13, 17, 19} son 4 => ((4+5)/20)
def primos_entre_decenas(veces):
    #10, 20, 30... (decenas)
    decenas = []
    #Fracciones finales
    fracciones = []
    #Primos requeridos
    all_primes = primos(veces*10)
    #5, 9, 11... (suma del número de primos por cada decena)
    decs = []
    #Ajusta las variables acorde con las veces(veces) requeridas
    i = 1
    while i<veces+1:
        decenas.append(i*10)
        decs.append(1)
        i += 1
    #Calcula las fracciones
    i = 0 
    while i<len(decenas):
        for n in all_primes:
            if n < decenas[i]:
                for a in range(1, len(decenas)+1):
                    if decenas[i] == multiplica(a, 10):
                        decs[i] += 1
        fracciones.append(decs[i]/decenas[i])
        i += 1
    debug('las fracciones son {}'.format(fracciones))
    return fracciones

#Devuelve la media de un array de números
def media(n_a):
    n_a_l = len(n_a)
    suma = 0
    i = 0
    while i<n_a_l:
        suma += float(n_a[i])
        i += 1
    res = suma / n_a_l
    debug('la media da {}'.format(res))
    return res

#GEOMETRÍA

#CÍRCULO
def perimetro_circulo(radio):
    return multiplica(radio, pi, 2)
def area_circulo(radio):
    return multiplica(expone(radio, 2), pi)

def circulo(radio):
    datos = {
            "radio": radio,
            "diametro": radio*2,
            "area": area_circulo(radio),
            "perimetro": perimetro_circulo(radio)
            }
    return datos

#CUADRADO
def perimetro_cuadrado(lado):
    return multiplica(lado, 4)
def area_cuadrado(lado):
    return expone(lado, 2)

def cuadrado(lado):
    datos = {
            "lado": lado,
            "area": area_cuadrado(lado),
            "perimetro": perimetro_cuadrado(lado)
            }
    return datos

#TRIÁNGULO EQUILÁTERO
def perimetro_triangulo_equilatero(lado):
    return multiplica(lado, 3)
def altura_triangulo_equilatero(lado):
    c = lado
    b = divide(lado, 2)
    a = math.sqrt(expone(c, 2) - expone(b, 2))
    return a
def area_triangulo_equilatero(lado):
    area = divide(multiplica(altura_triangulo_equilatero(lado), lado), 2)
    return area

def triangulo_equilatero(lado):
    datos = {
            "lado": lado,
            "area": area_triangulo_equilatero(lado),
            "perimetro": perimetro_triangulo_equilatero(lado)
            }
    return datos

#RECTÁNGULO
def perimetro_rectangulo(l1, l2):
    return suma(multiplica(l1, 2), multiplica(l2, 2))
def area_rectangulo(l1, l2):
    return multiplica(l1, l2)

def rectangulo(l1, l2):
    datos = {
            "lado1": l1,
            "lado2": l2,
            "perimetro": perimetro_rectangulo(l1, l2),
            "area": area_rectangulo(l1, l2)
            }
    return datos

#TRIÁNGULO ISÓCELES
def perimetro_triangulo_isoceles(lado, base):
    return suma(multiplica(lado, 2), base)
def altura_triangulo_isoceles(lado, base):
    c = lado;
    b = divide(base, 2)
    a = math.sqrt(expone(c, 2) - expone(b, 2))
    return a
def area_triangulo_isoceles(lado, base):
    altura = altura_triangulo_isoceles(lado, base)
    return divide(multiplica(base, altura), 2)

def triangulo_isoceles(lado, base):
    datos = {
            "lado": lado,
            "base": base,
            "altura": altura_triangulo_isoceles(lado, base),
            "perimetro": perimetro_triangulo_isoceles(lado, base),
            "area": area_triangulo_isoceles(lado, base)
            }
    return datos

#CUBO
def volumen_cubo(lado):
    return expone(lado, 3)
def superficie_cubo(lado):
    return multiplica(expone(lado, 2), 6)

def cubo(lado):
    datos = {
            "lado": lado,
            "volumen": volumen_cubo(lado),
            "superficie": superficie_cubo(lado)
            }
    return datos

#ESFERA
def volumen_esfera(radio):
    return multiplica(divide(4, 3), pi, expone(radio, 3))
def diametro_esfera(radio):
    return multiplica(radio, 2)
def ecuador_esfera(radio):
    return perimetro_circulo(radio)
def area_esfera(radio):
    return multiplica(4, pi, expone(radio, 2))

def esfera(radio):
    datos = {
            "radio": radio,
            "diametro": diametro_esfera(radio),
            "volumen": volumen_esfera(radio),
            "area": area_esfera(radio),
            "ecuador": ecuador_esfera(radio)
            }
    return datos


